compiler/lib/global_deadcode.ml

```diff
@@ -0,0 +1,435 @@
+        List.iter
+          ~f:(fun arg ->
+            match arg with
+            | Pv v -> add_use Compute x v
+            | Pc _ -> ())
+          args
+  in
+  Addr.Map.iter
+    (fun _ block ->
+      (* Add deps from block body *)
+      List.iter
+        ~f:(fun (i, _) ->
+          match i with
+          | Let (x, e) -> add_expr_uses x e
+          | Assign (x, y) -> add_use Compute x y
+          | Set_field (_, _, _) | Offset_ref (_, _) | Array_set (_, _, _) -> ())
+        block.body;
+      (* Add deps from block branch *)
+      match fst block.branch with
+      | Return _ | Raise _ | Stop -> ()
+      | Branch cont -> add_cont_deps cont
+      | Cond (_, cont1, cont2) ->
+          add_cont_deps cont1;
+          add_cont_deps cont2
+      | Switch (_, a) -> Array.iter ~f:add_cont_deps a
+      | Pushtrap (cont, _, cont_h, _) ->
+          add_cont_deps cont;
+          add_cont_deps cont_h
+      | Poptrap cont -> add_cont_deps cont)
+    prog.blocks;
+  uses
+
+(** Return the set of variables used in a given expression *)
+let expr_vars e =
+  let vars = Var.Set.empty in
+  match e with
+  | Apply { f; args; _ } ->
+      let vars = Var.Set.add f vars in
+      List.fold_left ~f:(fun acc x -> Var.Set.add x acc) ~init:vars args
+  | Block (_, params, _) ->
+      Array.fold_left ~f:(fun acc x -> Var.Set.add x acc) ~init:vars params
+  | Field (z, _) -> Var.Set.add z vars
+  | Prim (_, args) ->
+      List.fold_left
+        ~f:(fun acc v ->
+          match v with
+          | Pv v -> Var.Set.add v vars
+          | Pc _ -> acc)
+        ~init:vars
+        args
+  | Constant _ | Closure (_, _) -> vars
+
+(** Compute the initial liveness of each variable in the program. 
+
+    A variable [x] is marked as [Top] if 
+    + It is used in an impure expression (as defined by [pure_expr]);
+    + It is used in a conditonal/switch;
+    + It is raised by an exception;
+    + It is used in another stateful instruction (like setting a block or array field);
+    + Or, it is returned or applied to a function and the global flow analysis marked it as escaping.
+    
+    A variable [x[i]] is marked as [Live {i}] if it is used in an instruction where field [i] is referenced or set. *)
+let liveness nv prog pure_funs (global_info : Global_flow.info) =
+  let live_vars = Array.make nv Dead in
+  let add_top v =
+    let idx = Var.idx v in
+    live_vars.(idx) <- Top
+  in
+  let add_live v i =
+    let idx = Var.idx v in
+    match live_vars.(idx) with
+    | Live fields -> live_vars.(idx) <- Live (IntSet.add i fields)
+    | _ -> live_vars.(idx) <- Live (IntSet.singleton i)
+  in
+  let live_instruction i =
+    match i with
+    | Let (_x, e) -> (
+        if not (pure_expr pure_funs e)
+        then
+          match e with
+          | Apply { f; args; _ } ->
+              add_top f;
+              List.iter
+                ~f:(fun x -> if variable_may_escape x global_info then add_top x)
+                args
+          | _ ->
```

Comment:
Make the pattern exhaustive

Reply:
[5dfd1ca](https://github.com/ocsigen/js_of_ocaml/pull/1503/commits/5dfd1caed57d2d0c85606063347b77238cffdb77)